### ISSUE TITLE
Stack View async loading not announced by screen reader

### ISSUE THREAD

**Issue by gnomeontherun** [vmware/clarity#3565]
## Describe the bug

The "Loading..." message in the Lazy loading stack views example is not announced by screen readers.

## How to reproduce

https://stackblitz.com/github/clane/clarityTest2/tree/master/clarity-test-app

- Using a screen reader find the 'Lazy Loading of Children' example and activate one of the 'stack-block-label' buttons. 
- Observe that the text 'Loading...' appears
- Observe that the screen reader does not announce this text

## Expected behavior

It should announce loading.

## Versions

**App**

* Angular: [e.g. 6]
* Node: [e.g. 8.10.0]
* Clarity: [e.g. 0.12.5]

**Device:**

* Type: [e.g. MacBook]
* OS: [e.g. iOS]
* Browser [e.g. chrome, safari]
* Version [e.g. 22]

## Additional notes

See https://github.com/vmware/clarity/issues/3423 and the solutions applied elsewhere.

**Comment by andreiconstantinbytex** [comment]
I've looked into this and adding aria-label on the 'Loading...' div should fix this one. Can I be assigned to this one?

**Comment by Jinnie** [comment]
You are welcome :)

**Comment by andreiconstantinbytex** [comment]
Thank you!

**Comment by andreiconstantinbytex** [comment]
My opinion is that the lazily loaded stack view code example shouldn't have the aria-label set but I might be mistaken. Should it?

![image](https://user-images.githubusercontent.com/52167120/64510940-de493d00-d2eb-11e9-8712-8e342361c071.png)

**Comment by Jinnie** [comment]
@chlane can you comment on this, please?

**Comment by Jinnie** [comment]
Chris Lane 6:23 PM
     This issue is addressed in https://github.com/vmware/clarity/issues/3423

**Comment by chlane** [comment]
aria-label is necessary for cases where a UI component does not have a visible textual name. In this case we are talking about visible text i.e. <div>loading ....</div>, there is no need to label this. This is simply content that the screen reader user will hear when they eventually read it with their screen reader's cursor. But the sudden appearance of this message can be considered to be a status message, in which case a live region is the best solution. A live region will announce its contents regardless of the screen reader user reading another part of the document.  See https://github.com/vmware/clarity/issues/3423 for more information.

**Comment by gnomeontherun** [comment]
The aria-live region is a tricky thing for us from an API perspective, we need to be careful how this gets implemented. We have our new spinner component that handles the aria-live, which should be the recommendation for anything loading like this.

**Comment by andreiconstantinbytex** [comment]
As discussed with @gnomeontherun the 'Loading...' label will be replaced with the spinner.

**Comment by andreiconstantinbytex** [comment]
I've tried replacing the 'Loading...' label with the spinner but the screen reader still doesn't announce the fact that the list is loading. It says 'loading' only after you hover over the spinner. This is the exact same behavior even without the spinner. I've tried two different screen readers on two different OSs.

**Comment by chlane** [comment]
Hi Andrei,

What screen reader browser combinations are you using?

And are you using a mouse?

With the screen reader on, can you try using the Tab key to focus the desired button and the Enter key to activate it?

Chris

**Comment by andreiconstantinbytex** [comment]
So I used the following screen readers: Ubuntu 18.04 OS default screen reader, Win 10 OS Narrator + NVDAccess. No solution worked even with the keyboard approach.

**Comment by chlane** [comment]
Hi Andrei,

Clarity supports the Windows and macOS screen reader/browser combinations below.

•       Internet Explorer 10 or 11 and the latest version of JAWS or NVDA
•       Latest versions of Chrome, Firefox or Edge with the latest version of JAWS or NVDA
•       Latest version of Safari with the latest or previous version of VoiceOver

Can you try any of these screen reader/browser combinations?
Chris

**Comment by clane** [comment]
@gnomeontherun I noticed our current spinner implementation adds the live region dynamically. The WCAG approach would be to populate the live region dynamically rather than add it to the page dynamically.

Could we use *ngif to populate the live region rather than add it as a whole as seen below?

<clr-spinner *ngIf="downloadinInvoice" clrMedium clrAssertive>
<span *ngIf="downloadinInvoice">
Downloading
</span>
</clr-spinner>

**Comment by bdryanovski** [comment]
By default the `clrSpinner` will set `aria-live="polite"` if no other option is set to `true` for example `clrAssrtive` or `clrOff`.

So placing `clrSpinner` without `*ngIf` will make the screen reader announce it when the page renders finish in general case this will be when everything on the page is loaded. So we use the `*ngIf` directive to control when this happens and also to display the loaders animation.

@andreiconstantinbytex is there any demo page that I could take a look - the stackblitz from the first comment is a bit general and have to make some changes to it to test it - so I could make something different than you.

**Comment by clane** [comment]
Can we place an empty live region on the page like this

`<div role="status" clrAssrtive></div>`

then populate it with the Loading icon and text?

`<div role="status" clrAssrtive>
<clr-spinner>...</clr-spinner>
</div>`

**Comment by gnomeontherun** [comment]
The spinner needs to have `*ngIf` on it as well, or it won't announce. We decided against trying to setup a general purpose aria-live region.

**Comment by chlane** [comment]
I’m not proposing a global live region, but rather that we change our technique from insertion of a live region to population of a live region.

**Comment by gnomeontherun** [comment]
Ok, that is a separate bug that should be filed and should be done with the spinner everywhere. This can proceed as planned then and the other bug will fix the limited support issue.

**Comment by andreiconstantinbytex** [comment]
So inserting the spinner should suffice for now and then the bug with the screen reader will be handled separately, correct?

**Comment by gnomeontherun** [comment]
@andreiconstantinbytex correct!

**Comment by andreiconstantinbytex** [comment]
Thank you for the update.

**Comment by Jinnie** [comment]
Seems this was merged.
